Retry: Low mem max_pool2d_with_indices

amjames opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122832

Based on #105687

The low memory path does not need to strictly return the int8 offsets
instead the offset to index computation can be separated from the
inner function of the max pool lowering. The partitioner can then choose
to move the offset to index computation into the backward pass.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @desertfire @chauhang

amjames committed:
[WIP] Retry: Low mem max_pool2d_with_indices

Based on #105687

The low memory path does not need to strictly return the int8 offsets
instead the offset to index computation can be separated from the
inner function of the max pool lowering. The partitioner can then choose
to move the offset to index computation into the backward pass.

[ghstack-poisoned]

eellison commented:
Let me know when you want a review here.

eellison commented:
Request a review when/if you want me to take a look

amjames committed:
Update on "[WIP] Retry: Low mem max_pool2d_with_indices"


Based on #105687

The low memory path does not need to strictly return the int8 offsets
instead the offset to index computation can be separated from the
inner function of the max pool lowering. The partitioner can then choose
to move the offset to index computation into the backward pass.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler desertfire chauhang

[ghstack-poisoned]

peterbell10 commented on torch/_functorch/partitioners.py:
```diff
@@ -733,7 +733,7 @@ def classify_nodes(joint_module):
     view_ops = [aten.squeeze, aten.unsqueeze, aten.alias]
     if compiler == "inductor":
-        default_recomputable_ops += [prims.div, prims.convert_element_type, aten.clone, aten._to_copy, aten.full_like, prims.var, prims.sum, aten.var, aten.std, prims.broadcast_in_dim, aten.select, aten.permute, aten._unsafe_view, aten.view, aten.expand, aten.slice, aten.reshape, aten.broadcast_tensors, aten.scalar_tensor, aten.ones, aten.new_zeros, aten.lift_fresh_copy, aten.arange, aten.triu, a ...
+        default_recomputable_ops += [prims.div, prims.convert_element_type, aten.clone, aten._to_copy, aten.full_like, prims.var, prims.sum, aten.var, aten.std, prims.broadcast_in_dim, aten.select, aten.permute, aten._unsafe_view, aten.view, aten.expand, aten.slice, aten.reshape, aten.broadcast_tensors, aten.scalar_tensor, aten.ones, aten.new_zeros, aten.lift_fresh_copy, aten.arange, aten.triu, a ...
```
This should work if you add an 
```python
import torch._inductor.inductor_prims
```

eellison commented on torch/_inductor/lowering.py:
```diff
@@ -3912,20 +3926,115 @@ def fn(idx, return_index):
+    )
+    # If we landed here, the indices are not used for the backward pass, realize
+    # hint here so the conversion is fused into the forward loop.
+    indices.realize_hint()
```
Realization in inductor is separate from the forward and backward partitioning. First we do aot_autograd partitioning, then we compile each separately.

I don't think we need to realize the indices here. If they're an output of forward saved for backward they will get realized as graph outputs.

eellison commented on torch/_inductor/lowering.py:
```diff
@@ -3912,20 +3926,115 @@ def fn(idx, return_index):
+            assert input_width.is_integer
+            return ops.index_expr(input_width, torch.int64)
+        else:
+            # routing through adaptive_max_pool to here the shape is evaluated
+            # so we get a constant
+            return ops.constant(input_width, torch.int64)
```
Hmm, `routing through adaptive_max_pool to here the shape is evaluated` 

Hmm, what do you mean ?

amjames replied:
I was trying to explain why a branch was needed there, as looking at the paths through  `max_pool2d_with_indices` it seems it should not be needed.  Indeed we don't hit this branch though any of the `max_pool2d` tests. 


`adaptive_max_pool2d` will route through here under [special conditions](https://github.com/pytorch/pytorch/blob/c63a7b569133c9d91bde362c68e4f60abd4b619b/torch/_inductor/lowering.py#L4305C4-L4307C55). In that context the lowering for `adaptive_max_pool` has called `evaluate_static_shape` on the input sizes and `input_width` is always a constant.

eellison reviewed (commented):
Looking good , a good couple comments. also, oops, didn't mean to comment twice above, was just clearing review queue..

amjames committed:
Update on "Retry: Low mem max_pool2d_with_indices"


Based on #105687

The low memory path does not need to strictly return the int8 offsets
instead the offset to index computation can be separated from the
inner function of the max pool lowering. The partitioner can then choose
to move the offset to index computation into the backward pass.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler desertfire chauhang

[ghstack-poisoned]

amjames commented on test/inductor/test_torchinductor.py:
```diff
@@ -3556,7 +3556,7 @@ def fn(x):
 
         self.common(
             fn,
-            (-torch.arange(1 * 8 * 8, dtype=torch.float32).view(1, 1, 8, 8),),
```
This could also be made a skip for the dynamic codegen tests, but the outer loop over `batch x channels` specializes when that product is 1.

amjames committed:
Update on "Retry: Low mem max_pool2d_with_indices"


Based on #105687

The low memory path does not need to strictly return the int8 offsets
instead the offset to index computation can be separated from the
inner function of the max pool lowering. The partitioner can then choose
to move the offset to index computation into the backward pass.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler desertfire chauhang

[ghstack-poisoned]

amjames committed:
Update on "Retry: Low mem max_pool2d_with_indices"


Based on #105687

The low memory path does not need to strictly return the int8 offsets
instead the offset to index computation can be separated from the
inner function of the max pool lowering. The partitioner can then choose
to move the offset to index computation into the backward pass.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler desertfire chauhang

[ghstack-poisoned]

amjames committed:
Update on "Retry: Low mem max_pool2d_with_indices"


Based on #105687

The low memory path does not need to strictly return the int8 offsets
instead the offset to index computation can be separated from the
inner function of the max pool lowering. The partitioner can then choose
to move the offset to index computation into the backward pass.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler desertfire chauhang

[ghstack-poisoned]

amjames committed:
Update on "Retry: Low mem max_pool2d_with_indices"


Based on #105687

The low memory path does not need to strictly return the int8 offsets
instead the offset to index computation can be separated from the
inner function of the max pool lowering. The partitioner can then choose
to move the offset to index computation into the backward pass.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler desertfire chauhang

[ghstack-poisoned]

peterbell10 reviewed (changes requested):
The code looks good overall, minus a few nits. The compression ratio improvements look great, but some of the perf numbers are a bit concerning.

eellison commented on torch/_inductor/inductor_prims.py:
```diff
@@ -93,3 +93,36 @@ def eager_force_stride(input_tensor: Tensor, stride) -> Tensor:
+
+_low_memory_max_pool2d_offsets_to_indices = make_prim(
+    "_low_memory_max_pool2d_offsets_to_indices(Tensor self, SymInt kernel_h, SymInt input_w, SymInt[2] stride, SymInt[2] padding) -> Tensor",
+    lambda self, *args: self.to(torch.int64),
```
Do we want to change the dtype here dynamically ? typically we dont need int64 right ?

amjames replied:
I don't think that would be sound in general. It should be internal to the `offset_to_index->backward` fusion. There are tests which fail if output dtypes do not match eager.

eellison commented on torch/_inductor/decomposition.py:
```diff
@@ -705,3 +706,50 @@ def _softmax_backward_data(grad_output, output, dim, input_dtype):
+    # If the fallback is not being selected we route through the low-memory path.
+    # int8 offsets will be cached for backward pass.
+    # If the backward is lowered offset->index will be inlined.
+    vals, offsets = prims._low_memory_max_pool2d_with_offsets(
```
Looking at the kernel generated in `functorch_dp_cifar10`: https://gist.github.com/eellison/481ea321a5c46d006879f35154d5b508. 
 
We get speedup here by changing the int64 -> int32 in generated triton code. Arguably the value range analysis code should be updated to handle that. But at least in interim, can we update decomp/lowering such that we use int32 dtype when possible ?

eellison reviewed (commented):
Memory compression looks great ! And I like the decomposition approach as opposed to lowering - definitely simpler / more maintainable. 

I think just a couple of small things left per @peterbell10 and my comments.

peterbell10 commented on torch/_inductor/lowering.py:
```diff
@@ -3556,20 +3570,108 @@ def fn(idx, return_index):
+
+    def increments_to_index(h_inc, w_inc, bh, bw):
+        w_in = get_input_width_opsvalue()
+        stride_expr = [ops.constant(s, torch.int64) for s in stride]
+        padding_expr = [ops.constant(p, torch.int64) for p in padding]
+        ih = bh * stride_expr[0] + h_inc - padding_expr[0]
+        iw = bw * stride_expr[1] + w_inc - padding_expr[1]
```
I think this might incidentally fix the int64 indexing, since `ops.index_expr` does use the kernels index dtype for integer expressions.

amjames committed:
Update on "Retry: Low mem max_pool2d_with_indices"


Based on #105687

The low memory path does not need to strictly return the int8 offsets
instead the offset to index computation can be separated from the
inner function of the max pool lowering. The partitioner can then choose
to move the offset to index computation into the backward pass.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler desertfire chauhang

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/inductor_prims.py:
```diff
@@ -16,8 +16,15 @@ def make_prim(
+    if isinstance(return_type, tuple):
+
+        def meta(*args, **kwargs):
+            return tuple(_prims.TensorMeta(o) for o in impl_aten(*args, *kwargs))
```
```suggestion
            return tuple(_prims.TensorMeta(o) for o in impl_aten(*args, **kwargs))
```

peterbell10 commented on torch/_inductor/lowering.py:
```diff
@@ -3556,20 +3577,68 @@ def fn(idx, return_index):
+    return out, offsets
+
+
+@register_lowering(aten.max_pool2d_with_indices, type_promotion_kind=None)
+def max_pool2d_with_indices(
+    x, kernel_size, stride=None, padding=0, dilation=1, ceil_mode=False
+):
+    # assert we are on the fallback path, otherwise we should use the the
+    # prims._low_memory variant
+    kernel_size, stride, padding, dilation, _ = max_pool2d_checks(
+        x, kernel_size, stride, padding, dilation, assert_fallback=True
+    )
+    return fallback_max_pool2d_with_indices(
+        x, kernel_size, stride, padding, dilation, ceil_mode
+    )
```
Can you not use `make_fallback` here?

amjames replied:
Yes, I wanted the assert here temporarily to make sure I had covered paths like `adaptive_max_pool2d`.

amjames committed:
Update on "Retry: Low mem max_pool2d_with_indices"


Based on #105687

The low memory path does not need to strictly return the int8 offsets
instead the offset to index computation can be separated from the
inner function of the max pool lowering. The partitioner can then choose
to move the offset to index computation into the backward pass.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler desertfire chauhang

[ghstack-poisoned]

amjames committed:
Update on "Retry: Low mem max_pool2d_with_indices"


Based on #105687

The low memory path does not need to strictly return the int8 offsets
instead the offset to index computation can be separated from the
inner function of the max pool lowering. The partitioner can then choose
to move the offset to index computation into the backward pass.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler desertfire chauhang

[ghstack-poisoned]

peterbell10 commented:
@amjames have you rerun the benchmarks?

amjames commented:
@peterbell10 Yes, [looking much better](https://hud.pytorch.org/benchmark/torchbench/inductor_with_cudagraphs?startTime=Mon,%2015%20Apr%202024%2014:45:43%20GMT&stopTime=Mon,%2022%20Apr%202024%2014:45:43%20GMT&granularity=hour&mode=training&dtype=amp&lBranch=gh/amjames/14/orig&lCommit=de2adf79b0f5d50ac3844db187b507fba8d97a9f&rBranch=main&rCommit=704fac5618dbefcf6b29832c0215ae5f5e0fa2c3)

peterbell10 commented:
Did you check what's going on with `resnet152`?

eellison reviewed (approved):
Looks good! I think if the overall numbers improve (which they do) we shouldnt block landing on perf decrease for one particular model. Although it might be a good follow up to look into. cc @peterbell10 on other opinions

eellison reviewed (approved):
Looks good! I think if the overall numbers improve (which they do) we shouldnt block landing on perf decrease for one particular model. Although it might be a good follow up to look into. cc @peterbell10 on other opinions

peterbell10 commented on torch/_inductor/lowering.py:
```diff
@@ -3932,7 +3983,20 @@ def adaptive_max_pool2d(x, output_size):
         )
     if h_in % h_out == 0 and w_in % w_out == 0:
         kernel_size = [h_in // h_out, w_in // w_out]
-        return max_pool2d_with_indices(x, kernel_size)
+        if should_fallback_max_pool2d_with_indices(kernel_size, dilation=[1, 1]):
+            return max_pool2d_with_indices(x, kernel_size)  # type: ignore[name-defined]   # noqa: F821
+        else:
+            v, offsets = _low_memory_max_pool2d_with_offsets(
+                x,
+                kernel_size,
+                stride=kernel_size,
+                padding=[0, 0],
+                dilation=[1, 1],
+                ceil_mode=False,
+            )
+            indices = _low_memory_max_pool2d_offsets_to_indices(
+                offsets, kernel_size[1], w_in, kernel_size, padding=[0, 0]
+            )
```
This isn't returning in the non-fallback case? Also having the two kernels in the lowering doesn't actually allow for the offsets to be saved for backward.

I would suggest adding an inductor-specific decomposition for these special cases that calls `max_pool2d_with_indices`.

amjames replied:
Fixed the missing return.

> Also having the two kernels in the lowering doesn't actually allow for the offsets to be saved for backward.

Yes, that is true. There no improvement for `adaptive_max_pool2d` from this PR, but there is also no penalty. 

> I would suggest adding an inductor-specific decomposition for these special cases that calls `max_pool2d_with_indices`.

I did try to set up an inductor decomposition for this, but since the `kernel_size` is shape dependent on this path you still need to handle these special cases in the lowering.

Given the complexity this seemed to require I planned to address it in a follow up.

eellison replied:
When would we actually expect to not want to save indices ? I don't know if we need to fix this.

amjames replied:
I think @peterbell10 is referring to saving the int8 offsets and recomputing the int64 indices.  

When the split happens at this level (inside the lowering) the partitioner won't be able to see a call to `inductor_prims._low_memory_max_pool2d_offsets_to_indices` which it can recompute on the backward so the int64 indices are always saved for `adaptive_max_pool2d`. 

I agree that we don't need to fix this. It would be nice to at some point, but we dont need to now. I nductor has no problem fusing the "extra" `offsets_to_indices` kernel with the first one, so this effectively generates the same code as it did when there was a call to the old lowering for `max_pool2d_with_indices` here.

peterbell10 replied:
Sure we can do this in a follow up PR. I don't follow this point though:

> since the `kernel_size` is shape dependent on this path you still need to handle these special cases in the lowering.

You should be able to compute `kernel_size` from `x.shape` and `out_shape` in the same way the lowering does.

amjames replied:
Yes, but here we `evaluate_static_shape` beforehand. When this logic is 
 moved out to an inductor decomposition x.shape` is symbolic and evaluating the conditional will throw.

peterbell10 replied:
>  x.shape` is symbolic and evaluating the conditional will throw.

It should just add a guard, no?

eellison replied:
We add guards on height and width of max_pool anyway

peterbell10 commented on torch/_inductor/lowering.py:
```diff
@@ -3473,22 +3473,22 @@ def pooling_size(x, i, kernel_size, stride, padding, ceil_mode):
-    aten.max_pool2d_with_indices.default,
-    add_to_fallback_set=False,
-)
+def should_fallback_max_pool2d_with_indices(kernel_size, dilation):
```
`kw_const = ops.constant(kernel_width, torch.int32)` should be fine since `kernel_width` is at most 25. 

`hbase = ops.index_expr(bh * stride[0] - padding[0], torch.int32)` would in theory be a problem except that in practice `ops.index_expr` respects the kernel indexing dtype, or at least this is true for triton. That said, for the same reason we should be able to use `torch.int64` here without loss in perf.

amjames committed:
Update on "Retry: Low mem max_pool2d_with_indices"


Based on #105687

The low memory path does not need to strictly return the int8 offsets
instead the offset to index computation can be separated from the
inner function of the max pool lowering. The partitioner can then choose
to move the offset to index computation into the backward pass.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler desertfire chauhang

[ghstack-poisoned]

peterbell10 commented:
Was there a follow up to check `resnet152`?

amjames commented:
@peterbell10 I did take a look at it and I didn't see any regression on the kernels generated from `max_pool` or backward. It also looks like that model was removed/disabled on the dashboard?

peterbell10 reviewed (approved):
Okay, seems to just be noise in the benchmark. 

Would still like to see the `adaptive_max_pool2d` special case changed to a decomp as a follow-up.

amjames commented:
@pytorchbot merge -r

amjames committed:
Update

[ghstack-poisoned]

CaoE commented on torch/_inductor/lowering.py:
```diff
@@ -3511,22 +3511,22 @@ def pooling_size(x, i, kernel_size, stride, padding, ceil_mode):
+def should_fallback_max_pool2d_with_indices(kernel_size, dilation):
+    kernel_size = pad_listlike(kernel_size, 2)
+    window_size = kernel_size[0] * kernel_size[1]
+    return (window_size > 25) or any(d > 1 for d in dilation)
```
Hi @amjames May I know why the threshold set here is 25?

amjames replied:
Sorry I missed this, the window size limit is due to the fact that inductor unrolls loops. 

I refactored this lowering to add the small int path for indices and the logic for selecting the fallback was moved around, but I did not generate it. We are pretty consistent across pooling kernels though, so I suspect that going beyond 25 the overhead for compiling the unrolled statement becomes detrimental.
